uaclient/config.py

```diff
@@ -236,15 +239,49 @@ def _status(self) -> "Dict[str, Any]":
         return response
 
     def status(self) -> "Dict[str, Any]":
-        """Return status as a dict, using a cache for non-root users"""
-        if os.getuid() == 0:
-            status = self._status()
-            self.write_cache("status-cache", status)
-            return status
-        cached_status = cast("Dict[str, Any]", self.read_cache("status-cache"))
-        if not cached_status:
-            return DEFAULT_STATUS
-        return cached_status
+        """Return status as a dict, using a cache for non-root users
+
+        When unattached, get available resources from the contract service
+        to report detailed availability of different resources for this
+        machine.
+
+        Write the status-cache when called by root.
+        """
+        from uaclient.contract import get_available_resources
+        from uaclient.entitlements import ENTITLEMENT_CLASS_BY_NAME
+
+        if not self.is_attached:
+            response = copy.deepcopy(DEFAULT_STATUS)
+            resources = get_available_resources(self)
+            for resource in sorted(resources, key=lambda x: x["name"]):
+                if resource["available"]:
+                    available = status.UserFacingAvailability.AVAILABLE.value
+                else:
+                    available = status.UserFacingAvailability.UNAVAILABLE.value
+                ent_cls = ENTITLEMENT_CLASS_BY_NAME.get(resource["name"])
+                if not ent_cls:
+                    LOG.debug(
+                        "Ignoring availability of unknown service {name}"
+                        " from contract server".format(name=resource["name"])
+                    )
+                    continue
+                response["services"].append(
+                    {
+                        "name": resource["name"],
+                        "description": ent_cls.description,
+                        "available": available,
```

Comment:
I'm a little uneasy about the fact that this service dict has different members to the service dict that attached status produces, whilst living under the same key. I'm worried that it makes parsing our JSON output quite painful (if attached: <one parsing implementation> else: <completely different implementation>). Maybe this is fine because those are two very different modes of operation, but it's making my spidey sense tingle so I'm bringing it up.

What do you think?

Reply:
@OddBloke, I was uneasy about this too. Though I felt that even standardizing these keys would still require extra (is|is_not)_attached to determine whether we use 'available' vs. 'entitled' keys. But I suppose that 'logic' can be embedded in our template strings which would only contain {available} vs {entitled} subst vars.


Some alternatives to keep standardized service keys in both attached and unattached status:
  1.   a. Extend attached status to add service availability = None for each service
        b. Extend unattached status to also include service 'entitled' = None for each service
  2. a. Extend attached status to actually provide availability key values
          * incurs extra remote contract API cost because we have to get those values even though not needed for status
      b. Extend unattached status to include 'entitled' = n/a


Not sure whether we should pay the API cost of 2a given that the attached status shouldn't even be using this available key.

Reply:
Perhaps another alternative would be to not overload the services key? Attached status would continue to use the `services` key and unattached status would use another key (`resources`?). Then at least it's clear that they will contain different data structures, and consumers can use the top-level keys to determine what's going on (rather than discovering the differences when processing the first service entry).

A completely different alternative would be to continue with the code as-is, but include a message in help that the JSON output isn't yet considered stable. I'm leaning towards this option, as polluting the top-level namespace without careful thought (as I just suggested ^_^) feels likely to just be digging a deeper hole.

Reply:
I think I'd like to avoid adding another top-level key too, especially because we have formerly renamed 'entitlements' to 'services' and 'resources' are really just an alias for 'services'. So agreed, we'd probably dig a deeper hole adding a new key. I'll add a _doc key (which will surface at the top of the json response which will document that 'presence of keys under services are experimental' or something like that.

Reply:
We have this same _doc key in cloud-init instance-data.json.